src/libraries/System.Private.CoreLib/src/Microsoft/Win32/SafeHandles/SafeFileHandle.ThreadPoolValueTaskSource.cs

```diff
@@ -0,0 +1,187 @@
+
+                long result = 0;
+                try {
+                    bool notifyWhenUnblocked = false;
+                    try
+                    {
+                        // This is the operation's last chance to be cancelled.
+                        if (_cancellationToken.IsCancellationRequested)
+                        {
+                            _source.SetException(new OperationCanceledException(_cancellationToken));
+                            return;
+                        }
+
+                        notifyWhenUnblocked = ThreadPool.NotifyThreadBlocked();
+                        switch (_operation)
+                        {
+                            case Operation.Read:
+                                Memory<byte> writableSingleSegment = Unsafe.As<ReadOnlyMemory<byte>, Memory<byte>>(ref _singleSegment);
+                                result = RandomAccess.ReadAtOffset(_fileHandle, writableSingleSegment.Span, _fileOffset);
+                                break;
+                            case Operation.Write:
+                                result = RandomAccess.WriteAtOffset(_fileHandle, _singleSegment.Span, _fileOffset);
+                                break;
+                            case Operation.ReadScatter:
+                                Debug.Assert(_multiSegmentCollection != null && _multiSegmentCollection is IReadOnlyList<Memory<byte>>);
+                                result = RandomAccess.ReadScatterAtOffset(_fileHandle, (IReadOnlyList<Memory<byte>>) _multiSegmentCollection, _fileOffset);
+                                break;
+                            case Operation.WriteGather:
+                                Debug.Assert(_multiSegmentCollection != null && _multiSegmentCollection is IReadOnlyList<ReadOnlyMemory<byte>>);
+                                result = RandomAccess.WriteGatherAtOffset(_fileHandle, (IReadOnlyList<ReadOnlyMemory<byte>>)_multiSegmentCollection, _fileOffset);
+                                break;
+                        }
+                    } finally
+                    {
+                        if (notifyWhenUnblocked)
+                            ThreadPool.NotifyThreadUnblocked();
+                        _operation = Operation.None;
+                        _fileOffset = 0;
+                        _cancellationToken = default;
+                        _singleSegment = default;
+                        _multiSegmentCollection = null;
+                    }
+                } catch (Exception e)
+                {
+                    _source.SetException(e);
+                }
+                _source.SetResult(result);
+            }
+
+            public ValueTask<int> QueueRead(Memory<byte> buffer, long fileOffset, CancellationToken cancellationToken)
+            {
+                Debug.Assert(_operation == Operation.None, "An operation was queued before the previous one's completion.");
```

Comment:
Should the error message end with a period?

Reply:
This is fine; it's an assert, so no one should see it except a dev working on this code.

Reply:
That said, this is repeated several times.  Consider adding a method like:
```C#
[Conditional("DEBUG")]
private void ValidateInvariants()
{
    Operation op = _operation;
    Debug.Assert(op == Operation.None, "An operation was queued before the previous {op}'s completion.");
}
```
and then call that in each place rather than duplicating the assert.